### DISCUSSION TITLE
Possible Bug With Location-based Page Breaks?

### DISCUSSION THREAD

**Discussion by zicklag** [typst/typst#1827] [Q&A] [answered]
Hey everybody! I'm trying to use Typst to layout a book, but I'm struggling with something that feels like a bug, but might not be. I'd appreciate any help!

I'm trying to setup the page breaks so that the main chapters always start on the right-hand page. I think my code is sound, but it seems that typst get's confused with the locations of the chapter headings, and after a few pages, it gets out of sync with the actual page number.

Here's my code:

<details>

```typs
#let is-left-page(loc) = calc.even(loc.page())
#let is-right-page(loc) = not is-left-page(loc)

#let break-to-right() = locate(loc => {
  sym.zws
  if is-left-page(loc) {
    // Debug text:
    [[break-to-right #is-left-page(loc) #loc.page()]]
    pagebreak()
  }
})

#let table-of-contents() = locate(loc => {
  text(weight: "bold")[Contents]
  // Build the table cells for the contents listing
  let chapter-num = 1
  for chapter in query(<chapter-heading>, loc) { 
    [
      #grid(
        columns: (auto, 1fr, 10pt),
        [Chapter #chapter-num],
        align(center)[#link(chapter.location(), chapter)],
        align(right)[#chapter.location().page()],
      )
    ]

    // Increment chapter number
    chapter-num += 1
  }
})

#set page(
  paper: "a6",
  header: locate(loc => {
    align(center)[#loc.page()]
  })
)

// Modify chapter headings
#set heading(numbering: "1")
#show heading.where(level: 1): x => {
  pagebreak()
  break-to-right()

  v(20pt)
  let number = counter(heading).display(x.numbering)
  locate(loc => 
    [
      #text(1em, weight: "regular")[#smallcaps[Chapter] #number] \
      #text(1.5em, weight: 700, hyphenate: false)[#x.body <chapter-heading>] \
      Page: #loc.page()
    ]
  )
}

// Title
#v(5em)
#align(center, text(20pt)[My Book])

#table-of-contents()

= My Stuff

#lorem(200)

= Extra Info

#lorem(100)

= Another Chapter

#lorem(200)

= More Stuff

#text(20pt)[^ Wrong page number]

#lorem(500)

= It Just Keeps Comming

#lorem(80)

#lorem(50)

= Latin Stuff

#lorem(200)

#lorem(80)

#lorem(50)
```

</details>

Here's an example of the rendered PDF, showing that the location of the title is somehow mis-calculated, with the correct page being displayed in the page header, above it.

![image](https://github.com/typst/typst/assets/25393315/7f42eb5b-d4e7-4e8d-afe2-1f467702f505)

It ends up with the same issue for the `break-to-right()` function, too, where the page number is mis-calculated and it starts breaking pages so that the chapters start on the left page, instead of the right.

It seems like it might end up one page off, for every time that the conditional page break triggers, or something like that.

**Comment by Julian-Wassmann** [comment, marked as answer]
This can be done with the `pagebreak()` parameter `to: "even" | "odd"`. See the [documentation](https://typst.app/docs/reference/layout/pagebreak/#parameters-to). Use this instead of your custom `break-to-right()` function. Relying on page counters (states) and updating them at the same time is dangerous.

**Reply by zicklag** [reply]
Beautiful, thank you!
